Podstawowe operacje blogpost (#28)
akai-org/blog: content/posts/2021-12-10---Podstawowe-operacje.md

---
title: Podstawowe operacje w Javie
date: "2021-12-10"
template: "post"
draft: false
slug: "/posts/springcourse2/"
category: "back-end"
tags:
- "Java"
description: "Wstęp do podstawowych operacji w Javie"
---

![logo](/media/banner-java.png)

## Wstęp

> Język Java, jak każdy inny język programowania posiada pewne podstawowe narzędzia bez których pisanie kodu było by niemożliwe. Każdy początkujący programista musi nauczyć się kilku podstawowych mechanik dostępnych w każdym języku programowania. Są to **zmienne**, **pętle** oraz **instrukcje warunkowe**. W tym poście postaramy się omówić te trzy filary programowania na przykładzie Javy. Warto zaznaczyć, że narzędzia te są tak uniwersalne i wszechstronne, że zrozumienie ich na przykładzie Javy pozwoli Ci zastosować je w każdym innym języku.

Reviewer: ```suggestion
Java, jak każdy inny język programowania posiada pewne podstawowe narzędzia bez których pisanie kodu byłoby niemożliwe. Każdy początkujący programista musi nauczyć się kilku podstawowych mechanik dostępnych w każdym języku programowania. Są to **zmienne**, **pętle** oraz **instrukcje warunkowe**. W tym poście postaramy się omówić te trzy filary programowania na przykładzie Javy. Warto zaznaczyć, że narzędzia te są tak uniwersalne i wszechstronne, że zrozumienie ich na przykładzie Javy pozwoli Ci zastosować je w każdym innym języku.
```

- `Byte` - pojedynczy bajt (liczba całkowita 8-bitowa)
- `Short` - liczba całkowita 16-bitowa
- `Integer` - liczba całkowita 32-bitowa
- `Long` - liczba całkowita 64-bitowa
- `Float` - liczba zmiennoprzecinkowa
- `Double` - liczba zmiennoprzecinkowa o podwójnej precyzji
- `Boolean` - wartość logiczna (prawda/fałsz)
- `Character` - pojedynczy znak
- `String` - ciąg znaków

#### Autoboxing i unboxing
W rzeczywistości przykład deklaracji zmiennej typu *boxed* jest trochę bardziej skomplikowany. Na szczęście Java udostępnia mechanizm konwersji typu prostego na obiektowy (ang. *autoboxing*) i typu obiektowego na prosty (ang. *unboxing*). Poniżej przedstawię przykład działania tego jakże przydatnego mechanizmu.

```java
// implementacja typu prostego
int primitiveInt = 10;

// implementacja typu obiektowego
Integer boxedInt = new Integer(10);

// konwersja typu primitive na typ boxed (autoboxing)
Integer boxedInt = 10;

// konwersja typu boxed na typ primitive (unboxing)
int primitiveInt = Integer.valueOf(10);
```

### Pętle

> Pętle to kolejny mechanizm bez którego tworzenie oprogramowania było by nie możliwe. Pozwalają na wykonywanie tej samej części kodu wielokrotnie. Każde wykonanie kodu w ciele pętli i powrót do jej początku nazywamy iteracją pętli. Programista może z góry podać liczbę iteracji lub wykorzystać warunek logiczny aby określić, kiedy pętla powinna się zatrzymać.  

Reviewer: ```suggestion
Pętle to kolejny mechanizm bez którego tworzenie oprogramowania byłoby niemożliwe. Pozwalają na wykonywanie tej samej części kodu wielokrotnie. Każde wykonanie kodu w ciele pętli i powrót do jej początku nazywamy iteracją pętli. Programista może z góry podać liczbę iteracji lub wykorzystać warunek logiczny aby określić, kiedy pętla powinna się zatrzymać.  
```

W wielu językach programowania pętle wyglądają bardzo podobnie, wręcz identycznie. Dziś przyjrzymy się trzem rodzajom pętli, na które na pewno natkniesz lub natknąłeś/aś się podczas nauki programowania. Nie tylko w Javie :).

#### Pętla `for`

> Wspomniałem wcześniej, że programista może z góry założyć ilczbę iteracji pętli. `For` jest w tym przypadku nieoceniony. Deklaracja pętli `for` składa się z trzech części:

Reviewer: ```suggestion
Wspomniałem wcześniej, że programista może z góry założyć liczbę iteracji pętli. `For` jest w tym przypadku nieoceniony. Deklaracja pętli `for` składa się z trzech części:
```

```java
for (operacja1; operacja2; operacja3) {
  // ciało pętli
}
```
Po krótce przyjrzyjmy się tym częściom:
- *operacja1* - jest to kawałek kodu wykonywany na początku działania pętli, najczęściej definiujemy w nim *iterator* - specjalną zmienną służącą do liczenia iteracji pętli.
- *operacja2* - w tym miejscu deklarujemy warunek logiczny dzięki któremu opuścimy pętlę. Najczęściej jest to porównanie naszego *iteratora* do jakiejś zadanej wartości

> - *operacja3* - kawałek kodu wykonywany za każdą iteracją pętli. W tym miejscu możemy określić jak ma zmieniać się nasz *iterator* (na przykład zwiększać lub zmniejszać o jeden).  

Reviewer: ```suggestion
- *operacja3* - kawałek kodu wykonywany po każdej iteracji pętli. W tym miejscu możemy określić jak ma zmieniać się nasz *iterator* (na przykład zwiększać lub zmniejszać o jeden).  
```

}
```
Wyjście programu:
```
0
1
2
3
4
```
*Operacja1* deklaruje zmienną `i` typu `int` i ustawia jej wartość na `0`. *Operacja2* wskazuje nam, że pętla będzie się wykonywać dopóki wartość wcześniej zadeklarowanej zmiennej `i` będzie mniejsza niż `5`. *Operacja3* zwiększa nam wartość zmiennej `i` o jeden po każdej iteracji.  

Za pomocą pętli `for` możemy również stworzyć pętlę która będzie wykonywała się w nieskończoność:
```java
for (;;) {
  System.out.println("Nieskonczona petla!");
}
```

#### Specjalny przypadek pętli `for`: pętla `for-each`
`For-each` jest wykorzystaniem pętli `for` do iteracji po wszystkich elementach tablicy. Pętla ta działa tak, że podczas każdej iteracji specjalna zmienna przyjmuje wartości kolejnych elementów podanej tablicy. Deklaracja wygląda nastepująco:
```java
for (typZmiennej nazwaZmiennej : nazwaTablicy) {
  // ciało pętli
}
```
- *typZmiennej* jest typem elementów jakie występują w tablicy o nazwie *nazwaTablicy*. Jeżeli podajemy do pętli tablicę liczb naturalnych (`int`) to *typZmiennej* również musi wskazywać na `int`.  
- *nazwaZmiennej* wskazuje zmienną do której zapisywane będą kolejne elementy tablicy *nazwaTablicy*.

> - *nazwaTablicy* wskazuje tablicę z której kolejno odczytane elementy będą zapisane w zmiennej *nazwaZmiennej*.  

Reviewer: ```suggestion
- *nazwaTablicy* wskazuje obiekt będący tablicą lub impementujący interfejs `Iterable` z którego kolejno odczytane elementy będą zapisane w zmiennej *nazwaZmiennej*.  
```

Przykład użycia prezentuje się następująco:
```java
String[] przedmioty = {"polski", "angielski", "matematyka"};
for (String przedmiot : przedmioty) {
  System.out.println(przedmiot);
}
```
Wyjście programu:
```
polski
angielski
matematyka
```

#### Pętla `while`
Pętla `while` stanowi alternatywę dla pętli `for`. Cechuje się ona tylko jednym parametrem - warunkiem determinującym wyjście z pętli. W ten sposób programista nie musi deklarować iteratora tak jak w przydaku pętli `for`.  
Deklaracja pętli wygląda w sposób następujący:
```java
while (warunek) {
  // ciało pętli
}
```
Pętla wykonuje się tak długo, dopóki *warunek* pozostaje prawdziwy.

Przykład użycia pętli `while` wypisującej na konsolę kolejne cyfry od 0 do 4:
```java
int i = 0;
while (i < 5) {
  System.out.println(i);
  i++;
}
```
Wyjście programu:
```
0
1
2
3
4
```

W pętli `while` również możemy zaimplementować nieskończoną pętlę. Wystarczy w warunku pętli podać zawsze prawdziwą wartość (na przykład `true`). Przykład prezentuję poniżej.
```java
while (true) {
  System.out.println("Nieskonczona petla!");
}
```
#### Pętla `do-while`

> Pętla `do-while` jest bardzo podobna do pętli `while` z tą różnicą, że warunek wyjścia z pętli sprawdzany jest **po** wykonaniu ciała funkcji. Deklaracja zmiennej rozpoczyna się od słowa `do`, następnie podajemy ciało funkcji w klamrach, a na samym końcu instrukcje `while` z warunkiem.

Reviewer: ```suggestion
Pętla `do-while` jest bardzo podobna do pętli `while` z tą różnicą, że warunek wyjścia z pętli sprawdzany jest **po** wykonaniu ciała pętli. Gwarantuje to nam, że ciało pętli zostanie wykonane przynajmniej jeden raz. Deklaracja zmiennej rozpoczyna się od słowa `do`, następnie podajemy ciało pętli w klamrach, a na samym końcu instrukcje `while` z warunkiem.
```

```java
do {
  // ciało funkcji
} while (warunek);
```
Dla przykładu wykonamy naszą standardową już procedurę odliczania od 0 do 4:
```java
int i = 0;
do {
  System.out.println(i);
  i++;
} while (i < 5);
```
Pętla nieskończona wygląda bardzo podobnie do pętli `while`:
```java
do {
  System.out.println("Nieskonczona petla!");
} while (true);
```
### Instrukcje warunkowe

> Instrukcje warunkowe pozwalają nam na spawdzanie reguł logicznych i wykonywanie odpowiednich instrukcji w zależności od tego czy dana reguła jest prawdziwa czy nie. W skrócie moglibysmy określić ich działanie w ten sposób: *Jeżeli zaszło A to wykonaj B. W przeciwnym przypadku wykonaj C*.  

Reviewer: ```suggestion
Instrukcje warunkowe pozwalają nam na spawdzanie reguł logicznych i wykonywanie odpowiednich instrukcji w zależności od tego czy dana reguła jest prawdziwa czy nie. W skrócie moglibyśmy określić ich działanie w ten sposób: *Jeżeli zaszło A to wykonaj B. W przeciwnym przypadku wykonaj C*.  
```

Deklaracja pełnej instrukcji `switch-case` wygląda następująco:

```java
switch(wyrazenie) {
  case x:
    // kod wykonywany gdy zajdzie przypadek x
    break;
  case y:
    // kod wykonywany gdy zajdzie przypadek y
    break;
  default:
    // przypadek domyślny wykonywany gdy przypadki x oraz y nie zostaną dopasowane do wyrażenia
}
```
Zauważ, że po każdym przypadku na samym końcu umieszczamy instrukcję `break`. Gdybyśmy jej tam nie umieścili to po wykonaniu przypadku `x`, program mógłby wykonać również przypadek `y` lub przypadek domyślny.  
Przykład dzialania instrukcji `switch-case`:
```java
int dzienTygodnia = 4;
switch (dzienTygodnia) {
  case 1:
    System.out.println("Poniedziałek");
    break;
  case 2:
    System.out.println("Wtorek");
    break;
  case 3:
    System.out.println("Środa");
    break;
  case 4:
    System.out.println("Czwartek");
    break;
  case 5:
    System.out.println("Piątek");
    break;
  case 6:
    System.out.println("Sobota");
    break;
  case 7:
    System.out.println("Niedziela");
    break;
  default:
    System.out.println("Błąd! Dzień tygodnia musi być z przedziału 1-7!!!");
}
```
Wyjście programu:
```
Czwartek
```

> 

Reviewer: Jakeiś zakończenie by się przydało. Możesz coś wspomnieć o tym, że to koniec, ale powinniście wyczekiwać kolejnych części czy coś w tym stylu
